### ISSUE TITLE
JVM crash while serializing certain BoundingValueHierarchy objects on Windows

### ISSUE THREAD

**Issue by stephengold** [stephengold/Minie#41]
Prior discussion:  https://hub.jmonkeyengine.org/t/jvm-crash-in-stranded/47354

Reduced test app:
```java
    public void simpleInitApp() {
        assetManager.registerLocator(
                "C:\\Users\\sgold\\Downloads", FileLocator.class);
        Spatial geomPhy = assetManager.loadModel("pc-screen.gltf");

        CollisionShape meshShape
                = CollisionShapeFactory.createMeshShape(geomPhy);
        ChildCollisionShape[] children
                = ((CompoundCollisionShape) meshShape).listChildren();
        BoundingValueHierarchy bvh
                = ((MeshCollisionShape) (children[1].getShape())).getBvh();
        bvh.serialize();
        stop();
    }
```

Typical crash log (with an SpDebug native library):
```log
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffb954ae231, pid=12036, tid=13220
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.3+7 (17.0.3+7) (build 17.0.3+7)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.3+7 (17.0.3+7, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
# Problematic frame:
# C  [bulletjme.dll+0x27e231]
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  S U M M A R Y ------------

Command Line: -Dfile.encoding=windows-1252 -Duser.country=US -Duser.language=en -Duser.variant -ea mygame.Main

Host: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz, 12 cores, 31G,  Windows 11 , 64 bit Build 22621 (10.0.22621.3085)
Time: Thu Apr 18 12:42:45 2024 Pacific Daylight Time elapsed time: 2.874474 seconds (0d 0h 0m 2s)

---------------  T H R E A D  ---------------

Current thread (0x00000147381b2c70):  JavaThread "jME3 Main" [_thread_in_native, id=13220, stack(0x0000000377e00000,0x0000000377f00000)]

Stack: [0x0000000377e00000,0x0000000377f00000],  sp=0x0000000377eff0e0,  free space=1020k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [bulletjme.dll+0x27e231]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  com.jme3.bullet.collision.shapes.infos.BoundingValueHierarchy.serialize(J)[B+0
j  com.jme3.bullet.collision.shapes.infos.BoundingValueHierarchy.serialize()[B+6
j  mygame.Main.simpleInitApp()V+54
j  com.jme3.app.SimpleApplication.initialize()V+282
j  com.jme3.system.lwjgl.LwjglWindow.initInThread()Z+112
j  com.jme3.system.lwjgl.LwjglWindow.run()V+34
j  java.lang.Thread.run()V+11 java.base@17.0.3
v  ~StubRoutines::call_stub

siginfo: EXCEPTION_ACCESS_VIOLATION (0xc0000005), reading address 0xffffffffffffffff


Register to memory mapping:

RIP=0x00007ffb954ae231 bulletjme.dll
RAX=0xdddddddddddddddd is an unknown value
RBX={method} {0x000001473fdc08c0} 'serialize' '(J)[B' in 'com/jme3/bullet/collision/shapes/infos/BoundingValueHierarchy'
RCX=0x0000014739403a10 points into unknown readable memory: 0xdddddddddddddddd | dd dd dd dd dd dd dd dd
RDX=0x0000014741c78080 points into unknown readable memory: 0x0000000000000000 | 00 00 00 00 00 00 00 00
RSP=0x0000000377eff0e0 is pointing into the stack for thread: 0x00000147381b2c70
RBP=0x0000000377eff1c0 is pointing into the stack for thread: 0x00000147381b2c70
RSI=0x0000000000000002 is an unknown value
RDI=0x0 is NULL
R8 =0x0000000099999a70 is an unknown value
R9 =0x0000014741c78001 points into unknown readable memory: 01 2b 08 47 01 00 00
R10=0x00007ffb95230000 bulletjme.dll
R11=0x0 is NULL
R12=0x0 is NULL
R13={method} {0x000001473fdc08c0} 'serialize' '(J)[B' in 'com/jme3/bullet/collision/shapes/infos/BoundingValueHierarchy'
R14=0x0000000377eff1e8 is pointing into the stack for thread: 0x00000147381b2c70
R15=0x00000147381b2c70 is a thread


Registers:
RAX=0xdddddddddddddddd, RBX=0x000001473fdc08b8, RCX=0x0000014739403a10, RDX=0x0000014741c78080
RSP=0x0000000377eff0e0, RBP=0x0000000377eff1c0, RSI=0x0000000000000002, RDI=0x0000000000000000
R8 =0x0000000099999a70, R9 =0x0000014741c78001, R10=0x00007ffb95230000, R11=0x0000000000000000
R12=0x0000000000000000, R13=0x000001473fdc08b8, R14=0x0000000377eff1e8, R15=0x00000147381b2c70
RIP=0x00007ffb954ae231, EFLAGS=0x0000000000010206

Top of Stack: (sp=0x0000000377eff0e0)
0x0000000377eff0e0:   0000000099999a70 0000000000000010
0x0000000377eff0f0:   000001473fdc08b8 00000147381b2c70
0x0000000377eff100:   0000000000000000 00000147120ed48e
0x0000000377eff110:   99999a703fdc08b8 0000014739403a10
0x0000000377eff120:   0000014741c78080 0000000000000000
0x0000000377eff130:   000001473fdc08b8 00000147120ed621
0x0000000377eff140:   00000147381b2f18 0000000377eff1d0
0x0000000377eff150:   0000014739403a10 0000000000000002
0x0000000377eff160:   000001473fd3a4d8 0000000000000000
0x0000000377eff170:   00000147120ed2f6 0000000377eff178
0x0000000377eff180:   000001473fdc08b8 0000000377eff1e8
0x0000000377eff190:   000001473fdc0aa8 0000000000000000
0x0000000377eff1a0:   00000006223f4158 000001473fdc08b8
0x0000000377eff1b0:   0000000000000000 0000000377eff1e0
0x0000000377eff1c0:   0000000377eff238 00000147120e9657
0x0000000377eff1d0:   00000006223f4158 00000147120eaf7b 

Instructions: (pc=0x00007ffb954ae231)
0x00007ffb954ae131:   ec 38 48 8b 44 24 50 48 89 44 24 20 48 83 7c 24
0x00007ffb954ae141:   20 00 75 1c 4c 8d 05 14 f6 35 00 48 8b 15 bd eb
0x00007ffb954ae151:   43 00 48 8b 4c 24 40 e8 e6 e5 d8 ff 33 c0 eb 3f
0x00007ffb954ae161:   48 8b 4c 24 20 e8 87 50 d8 ff 83 f8 15 74 1c 4c
0x00007ffb954ae171:   8d 05 19 f6 35 00 48 8b 15 7a ec 43 00 48 8b 4c
0x00007ffb954ae181:   24 40 e8 bb e5 d8 ff 33 c0 eb 14 48 8b 4c 24 20
0x00007ffb954ae191:   e8 38 ab d8 ff 48 89 44 24 28 48 8b 44 24 28 48
0x00007ffb954ae1a1:   83 c4 38 c3 cc cc cc cc cc cc cc cc cc cc cc 4c
0x00007ffb954ae1b1:   89 44 24 18 48 89 54 24 10 48 89 4c 24 08 48 83
0x00007ffb954ae1c1:   ec 58 48 8b 44 24 70 48 89 44 24 38 48 83 7c 24
0x00007ffb954ae1d1:   38 00 75 1f 4c 8d 05 fc f5 35 00 48 8b 15 2d eb
0x00007ffb954ae1e1:   43 00 48 8b 4c 24 60 e8 56 e5 d8 ff 33 c0 e9 d6
0x00007ffb954ae1f1:   00 00 00 48 8b 4c 24 38 e8 3b 62 d9 ff 89 44 24
0x00007ffb954ae201:   34 8b 44 24 34 ba 10 00 00 00 8b c8 e8 99 6d d9
0x00007ffb954ae211:   ff 48 89 44 24 40 48 8b 44 24 38 48 8b 00 41 b1
0x00007ffb954ae221:   01 44 8b 44 24 34 48 8b 54 24 40 48 8b 4c 24 38
0x00007ffb954ae231:   ff 50 10 88 44 24 30 0f b6 44 24 30 85 c0 75 1c
0x00007ffb954ae241:   4c 8d 05 b8 f5 35 00 48 8b 15 a9 eb 43 00 48 8b
0x00007ffb954ae251:   4c 24 60 e8 ea e4 d8 ff 33 c0 eb 6d 8b 54 24 34
0x00007ffb954ae261:   48 8b 4c 24 60 e8 8b 22 d9 ff 48 89 44 24 48 48
0x00007ffb954ae271:   8b 4c 24 60 e8 50 3b d8 ff 0f b6 c0 85 c0 74 04
0x00007ffb954ae281:   33 c0 eb 45 48 8b 44 24 40 48 89 44 24 20 44 8b
0x00007ffb954ae291:   4c 24 34 45 33 c0 48 8b 54 24 48 48 8b 4c 24 60
0x00007ffb954ae2a1:   e8 f9 a1 d8 ff 48 8b 4c 24 60 e8 1a 3b d8 ff 0f
0x00007ffb954ae2b1:   b6 c0 85 c0 74 04 33 c0 eb 0f 48 8b 4c 24 40 e8
0x00007ffb954ae2c1:   b5 4d d9 ff 48 8b 44 24 48 48 83 c4 58 c3 cc cc
```

**Comment by stephengold** [comment]
I've confirmed that the issue is not present (for "pc-screen.gltf") on Linux.

Here's a crude disassembly of Windows64DebugSp_bulletjme.dll:
   18027e1b0:	4c 89 44 24 18       	mov    %r8,0x18(%rsp)
   18027e1b5:	48 89 54 24 10       	mov    %rdx,0x10(%rsp)
   18027e1ba:	48 89 4c 24 08       	mov    %rcx,0x8(%rsp)
   18027e1bf:	48 83 ec 58          	sub    $0x58,%rsp
   18027e1c3:	48 8b 44 24 70       	mov    0x70(%rsp),%rax
   18027e1c8:	48 89 44 24 38       	mov    %rax,0x38(%rsp)
   18027e1cd:	48 83 7c 24 38 00    	cmpq   $0x0,0x38(%rsp)
   18027e1d3:	75 1f                	jne    0x18027e1f4
   18027e1d5:	4c 8d 05 fc f5 35 00 	lea    0x35f5fc(%rip),%r8        # 0x1805dd7d8
   18027e1dc:	48 8b 15 2d eb 43 00 	mov    0x43eb2d(%rip),%rdx        # 0x1806bcd10
   18027e1e3:	48 8b 4c 24 60       	mov    0x60(%rsp),%rcx
   18027e1e8:	e8 56 e5 d8 ff       	call   0x18000c743
   18027e1ed:	33 c0                	xor    %eax,%eax
   18027e1ef:	e9 d6 00 00 00       	jmp    0x18027e2ca
   18027e1f4:	48 8b 4c 24 38       	mov    0x38(%rsp),%rcx
   18027e1f9:	e8 3b 62 d9 ff       	call   0x180014439
   18027e1fe:	89 44 24 34          	mov    %eax,0x34(%rsp)
   18027e202:	8b 44 24 34          	mov    0x34(%rsp),%eax
   18027e206:	ba 10 00 00 00       	mov    $0x10,%edx
   18027e20b:	8b c8                	mov    %eax,%ecx
   18027e20d:	e8 99 6d d9 ff       	call   0x180014fab
   18027e212:	48 89 44 24 40       	mov    %rax,0x40(%rsp)
   18027e217:	48 8b 44 24 38       	mov    0x38(%rsp),%rax
   18027e21c:	48 8b 00             	mov    (%rax),%rax
   18027e21f:	41 b1 01             	mov    $0x1,%r9b
   18027e222:	44 8b 44 24 34       	mov    0x34(%rsp),%r8d
   18027e227:	48 8b 54 24 40       	mov    0x40(%rsp),%rdx
   18027e22c:	48 8b 4c 24 38       	mov    0x38(%rsp),%rcx
   18027e231:	ff 50 10             	call   *0x10(%rax)                ;; crash occurred here

**Comment by stephengold** [comment]
I also confirmed that the issue is not present (for "pc-screen.gltf") on macOS.

**Comment by stephengold** [comment]
Testing old libraries to determine whether this is a regression...
(I was pleasantly surprised to discover that jitpack.io still has some old versions of Minie that never made it to MavenCentral!)

This issue is observed in Minie v2.0.0-test1 (Libbulletjme version 8.1.0), but not in Minie v1.7.0 (Libbulletjme v6.4.0).
Studying the relevant release notes, I found a change that looks highly relevant:
```quote
## Version 8.0.0 released on 10 June 2020
...
+ Split off `BoundingValueHierarchy` from the `MeshCollisionShape` class. (API change)
```

**Comment by stephengold** [comment]
Here's the suspicious commit:  https://github.com/stephengold/Libbulletjme/commit/2b3229486853c4a7ab38b541fa67c381c6b9f507

**Comment by stephengold** [comment]
I studied the 2b32294 commit, but didn't find anything to explain this issue. I did note a portability issue (always passing `true` for `i_swapEndian`) which predated the commit and would be masked by the platform checks in Minie.

Reading the comments in "btQuantizedBvh.h", on the other hand, is a sickening experience:
https://github.com/stephengold/Libbulletjme/blob/955e375aae6c59a74353720e47cbee97c69d52f3/src/main/native/bullet3/BulletCollision/BroadphaseCollision/btQuantizedBvh.h#L94-L130

The author of those comments clearly assumed that an `int` is 4 bytes wide and a `btVector3` is 12 bytes. `btQuantizedBvh` needs to be rewritten to deal with 2-byte `int`s and double-precision vectors.

**Comment by stephengold** [comment]
Note that Libbulletjme invokes `btOptimizedBvh::calculateSerializeBufferSize()`:

https://github.com/stephengold/Libbulletjme/blob/3aa04542019ad612facdff1aa312619ab14abead/src/main/native/glue/com_jme3_bullet_collision_shapes_infos_BoundingValueHierarchy.cpp#L99

while Bullet makes a point of using `calculateSerializeBufferSizeNew()` instead:

https://github.com/stephengold/Libbulletjme/blob/3aa04542019ad612facdff1aa312619ab14abead/src/main/native/bullet3/BulletCollision/CollisionShapes/btBvhTriangleMeshShape.cpp#L446

Maybe the serialization code Libbulletjme is using is obsolete. But at this point, it's probably easier to patch it up than to switch.

**Comment by stephengold** [comment]
I noticed another issue with BoundingValueHierarchy.cpp:  due to casting, the pointer passed to `btAlignedFree()` (for tracked NPO) might not be the same as the one that was acquired from `btAlignedAlloc()`. That could cause all sorts of subtle bugs. A cleaner solution would be to keep 2 pointers: one to the `btOptimizedBvh` and one to the allocated buffer.

Also, it would be wise to add some sanity checks the `deSerialize()` method (in debug builds, at least) such as making sure the buffer size, Bullet version, `m_useQuantization` flag, and such have the expected values.

**Comment by stephengold** [comment]
More worries:

`btBvhTriangleMeshShape::setLocalScaling()` may cause the `btOptimizedBvh` to be freed, reallocated, and rebuilt. This suggests that a BVH is only valid for a particular local scaling. This also creates complications for for `MeshCollisionShape.createShape()`, which wants to set the local scaling without necessarily computing the BVH.

**Comment by stephengold** [comment]
Also, `MeshCollisionShape::setOptimizedBvh()` should pass the scale factors to `btBvhTriangleMeshShape::setOptimizedBvh()`.

**Comment by stephengold** [comment]
Also, BoundingValueHierarchy_finalizeNative needs to invoke the `btOptimizedBvh` destructor.

**Comment by stephengold** [comment]
The native-code portion of the fix was committed at https://github.com/stephengold/Libbulletjme/commit/aff3b22703849ec74070c11dcb5b76187b45413b

After the next Libbulletjme release, this issue can probably be closed.

**Comment by stephengold** [comment]
I believe this issue is solved as of 1cc19d6d.

In hindsight, creating a separate Java object for each BVH (https://github.com/stephengold/Libbulletjme/commit/2b3229486853c4a7ab38b541fa67c381c6b9f507) was a risky change. `btBvhTriangleMeshShape` rebuilds its BVH for many reasons, making it difficult for Java and native code to stay in sync.

**Comment by stephengold** [comment]
I'll file new issue(s) for the on-going problems with loading/cloning `BoundingValueHierarchy`.
